### PR TITLE
python-support

### PR DESCRIPTION
**Pull request by lusentis**
- [x] python parser to extract functions and configuration (#139)
- [x] **libs/language-python/describeApi.js** wrapper code for the python parser
- [x] **libs/language-python/compile.js** compile (?) and copy all the files to .dawson-dist
- [x] **libs/language-python/installDeps.js** install dependencies
- [x] **libs/language-python/createIndex.js** create python-specific function wrapping logic (for Lambda)
- [x] [**commands/proxy.js**](https://github.com/dawson-org/dawson-cli/blob/python-support/src/commands/proxy.js) pass a `dockerImage` option (see below)
- [ ] [**commands/proxy.js**](https://github.com/dawson-org/dawson-cli/blob/python-support/src/commands/proxy.js) use docker-lambda to call authorizers too (@lusentis)
- [ ] update the devInstrument logic to work in request-reponse mode too, see: https://github.com/dawson-org/dawson-cli/commit/b71f52592f1748b8d672d6a143877fd676617eb2#diff-ef0939c2350c13f64e5e07d4865e833d
- [ ] [documentation](https://github.com/dawson-org/dawson-cli/blob/master/docs/README.md)

Example app: https://github.com/dawson-org/dawson-examples/tree/master/2-basic-python

### CODE REVIEW

**Comment by lusentis** [comment]
* I've not yet updated createIndex.spec.js to match the new changes

**Comment by lusentis** [comment]
Great work, thanks!
I'll update the test script in the package.json to include `cd src/python && pip install -r requirements-dev.txt && bash runtests.sh` :)

There's also a "global" App Configuration ([docs](https://dawson.sh/docs.html#5-application-configuration)), that is currently read from the `package.json`. Do you think it's OK to also require python users to have a package json in their repo? Or we may discuss another way... (currently we use the `name` and `dawson` properties).

File: src/python/runtests.sh [hunk 1]
```diff
@@ -0,0 +1 @@
+nosetests --with-coverage --verbosity=2
```
Comment (lusentis) [inline src/python/runtests.sh]: thank you :)

**Comment by lusentis** [comment]
@alexcasalboni tests are passing on this branch, if you can, rebase your fork :)

**Comment by alexcasalboni** [comment]
@lusentis will do! I also wanted to improve tests and make sure the JSON printed to stdout is always valid, so that you can proceed with the whole pipeline in JS without too much hacking. Then I'll need to work on the Python wrapper too.

Would you prefer making `createIndex.getWrappingCode` a bit more generic to handle every possible case, or creating a new `createIndex.getWrappingPythonCode` function? How will we choose which one to call? The language/runtime choice should be available along the pipeline.

**Comment by lusentis** [comment]
Great!

I'd propose refactoring the current implementation like this:

- [ ] `createIndex.js` should export a function which takes, in addition to the current parameters, the language/runtime (here: https://github.com/dawson-org/dawson-cli/blob/2a6c0ff88f3deb80f4930abdb0cf48954cef6f73/src/libs/createBundle.js#L181)
- [ ] `createIndex.js` should then call the appropriate function, which I'd put in a separate file, maybe under a `libs/language-python` folder
- [ ] The JS wrapping/packing logic could then be moved to a `libs/language-nodejs-latest` (effectively moving to a *'language-agnostic* dawson) - babel, npm install routines etc would go there

**Comment by lusentis** [comment]
Okay, I've ended up with a slightly different and more clean approach.
I've moved `createIndex` to `language-javascript-latest' and kept the whole language choice logic in createBundle.js (and config.js), so it will be much easier to add more languages in the future.

Feedback is appreciated :D

File: src/libs/createBundle.js [hunk 2]
```diff
@@ -131,6 +92,12 @@ async function uploadS3 ({ bucketName, uuid, tempZipFile, tempZipFileSize }) {
   return { zipS3Location };
 }
 
+function getFileExtension (language) {
+  if (language === LANGUAGE_JS_LATEST) {
```
Comment (alexcasalboni) [inline src/libs/createBundle.js]: I prefer switch cases :)

**Comment by alexcasalboni** [comment]
It looks good, so far! 

Regarding mixed-language projects, it would be cool! We should be able to easily detect both api.js and api.py, and parse/generate both on deploy.

File: src/libs/createBundle.js (hunk 2 above)
Comment (lusentis) [inline src/libs/createBundle.js]: done 👍

**Comment by lusentis** [comment]
Okay, I've just discovered that docker-lambda uses different docker images for different runtimes, so we must pass a `dockerImage` option to `dockerLambda` ([reference](https://github.com/lambci/docker-lambda/search?utf8=%E2%9C%93&q=dockerImage), [our code](https://github.com/dawson-org/dawson-cli/blob/python-support/src/commands/proxy.js#L171))

**Comment by alexcasalboni** [comment]
Awesome!

But wait.. does it mean that we can't have both at the same time in dev?

**Comment by lusentis** [comment]
no :) currently docker-lambda is invoked once for every function execution
and it does no container reuse, so we just specify the docker-lambda image
to use for each execution.
This is really easy to implement because we determine the image name from
the "api.runtime" property!

**Comment by lusentis** [comment]
I've cleaned up this branch a little:
- all the commits not strictly related to the python implementation have been moved to a branch named [`internal-language-refactor`](https://github.com/dawson-org/dawson-cli/tree/internal-language-refactor)
- `internal-language-refactor` has been merged into `master`
- this branch has been rebased onto master

🎉

**Comment by alexcasalboni** [comment]
Awesome. I've just opened a new PR to add the runtime config param.

**Comment by lusentis** [comment]
@alexcasalboni would you prefer to have write access to this branch?

**Comment by alexcasalboni** [comment]
Yeah, that would make things faster :)

**Comment by lusentis** [comment]
Done! (link: https://github.com/dawson-org/dawson-cli/invitations)

**Comment by lusentis** [comment]
I've just pushed a basic language-python implementation and I'm able to correctly deploy python functions 😄 
There's still some work to do (clean up, mostly)

Here's a py example I'm using: https://github.com/dawson-org/dawson-examples/tree/master/2-basic-python

File: src/libs/language-python/createIndex.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        return '"pong__"'
+
+      try:
+        result = ${getRunnerCode(name, apiConfig)}
+
+        ${hasEndpoint
+          ? `
+        # prepare response for api-gateway and auto stringify JSON if responseContentType is application/json
+        if event.get('meta') and 'application/json' in event.get('meta').get('expectedResponseContentType'):
+          return callback(null, { 'response': json.dumps(result) })
```
Comment (alexcasalboni) [inline src/libs/language-python/createIndex.js]: This will probably explode :)

File: src/libs/language-python/createIndex.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+      except Exception, err:
+        try:
+          # Promise rejections should be Errors containing a JSON-stringified 'message property'
+          # which contains the error information to be displayed.
+          # If the property is not valid JSON, the error is not exposed to the client
+          # and a generic HTTP 500 error will be exposed
+          json.loads(err.message);
+          print 'Lambda will terminate with error', err.message
+          raise Exception(err.message)
```
Comment (alexcasalboni) [inline src/libs/language-python/createIndex.js]: I suggest just a re-raise (to keep the original stacktrace):

```
raise
```

File: src/libs/language-python/createIndex.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        return {
+          'response': result
+        }
+        `
+        : `
+        # else, if this function has no "path" configured, we return the value as-is
+        return result
+        `}
+
+      except Exception, err:
```
Comment (alexcasalboni) [inline src/libs/language-python/createIndex.js]: I'd recommend the new syntax:

```
except Exception as err:
```

File: src/libs/language-python/createIndex.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        try:
+          # Promise rejections should be Errors containing a JSON-stringified 'message property'
+          # which contains the error information to be displayed.
+          # If the property is not valid JSON, the error is not exposed to the client
+          # and a generic HTTP 500 error will be exposed
+          json.loads(err.message);
+          print 'Lambda will terminate with error', err.message
+          raise Exception(err.message)
+        except ValueError, jsonErr:
```
Comment (alexcasalboni) [inline src/libs/language-python/createIndex.js]: same here (as jsonErr)

File: src/libs/language-python/createIndex.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+          print 'Unhandled error will be swallowed and reported as HTTP 500:'
+          print err
```
Comment (alexcasalboni) [inline src/libs/language-python/createIndex.js]: very minor:

```
print "%r" % err
```

File: src/libs/language-python/createIndex.js (hunk 3 above)
Comment (lusentis) [inline src/libs/language-python/createIndex.js]: ahah oops

Comment (lusentis) [inline src/libs/language-python/createIndex.js]: fixed in cbd4cae

File: src/libs/language-python/createIndex.js (hunk 4 above)
Comment (lusentis) [inline src/libs/language-python/createIndex.js]: updated in 7b34cd8

File: src/libs/language-python/createIndex.js (hunk 5 above)
Comment (lusentis) [inline src/libs/language-python/createIndex.js]: updated in 7b34cd8

File: src/libs/language-python/createIndex.js (hunk 6 above)
Comment (lusentis) [inline src/libs/language-python/createIndex.js]: updated in 7b34cd8

File: src/libs/language-python/createIndex.js (hunk 7 above)
Comment (lusentis) [inline src/libs/language-python/createIndex.js]: updated in 7b34cd8

**Comment by lusentis** [comment]
`dawson dev` now works too 🎉 (except for authorizers)
@alexcasalboni feel free to update any code that needs to be improved etc (I'm not too familiar with python in particular)

File: src/libs/language-python/api-parser/tests/modules/not_serializable.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    print(event, context)
+    return "Hello world"
+
+def another_function():
+    """ Just a normal function """
+    pass
+
+# this configuration is not JSON-serializable
+myfunction.api = {
+    "path": "/test",
```
Comment (lusentis) [inline src/libs/language-python/api-parser/tests/modules/not_serializable.py]: small change needed here: path should not start with a slash (see https://dawson.sh/docs.html#4-function-configuration)

File: src/libs/language-python/api-parser/tests/modules/one_dawson_class.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        return json.dumps(obj)
+
+        Which will return the following JSON:
+
+        {"text": "Hello World"}
+
+    """
+
+    api = {
+        "path": "/test"
```
Comment (lusentis) [inline src/libs/language-python/api-parser/tests/modules/one_dawson_class.py]: small change needed here: path should not start with a slash (see https://dawson.sh/docs.html#4-function-configuration)

File: src/libs/language-python/api-parser/tests/modules/one_dawson_function.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+""" A module with a dawson-valid Python function """
+
+def myfunction(event, context):
+    """ This function is dawson-valid """
+    print(event, context)
+    return "Hello world"
+
+myfunction.api = {
+    "path": "/test"
```
Comment (lusentis) [inline src/libs/language-python/api-parser/tests/modules/one_dawson_function.py]: small change needed here: path should not start with a slash (see https://dawson.sh/docs.html#4-function-configuration)

File: src/libs/language-python/api-parser/tests/modules/two_dawson_functions.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+""" A module with two dawson-valid Python functions """
+
+def myfunction(event, context):
+    """ This function is dawson-valid """
+    print(event, context)
+    return "Hello Function"
+
+myfunction.api = {
+    "path": "/test"
```
Comment (lusentis) [inline src/libs/language-python/api-parser/tests/modules/two_dawson_functions.py]: small change needed here: path should not start with a slash (see https://dawson.sh/docs.html#4-function-configuration)

File: src/libs/language-python/api-parser/tests/modules/two_dawson_functions.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+def myfunction2(event, context):
+    """ This function is dawson-valid """
+    print(event, context)
+    return "Hello world 2"
+
+myfunction2.api = {
+    "path": "/test2"
```
Comment (lusentis) [inline src/libs/language-python/api-parser/tests/modules/two_dawson_functions.py]: small change needed here: path should not start with a slash (see https://dawson.sh/docs.html#4-function-configuration)

**Comment by alexcasalboni** [comment]
Ok, great! I will have another look around and fix those paths asap.

File: src/config.js [hunk 13]
```diff
@@ -322,6 +326,11 @@ function describeApi (rootDir) {
         language: LANGUAGE_JS_LATEST,
         requiredApi: jsDescribeApi({ rootDir })
       };
+    } else if (existsSync(path.join(rootDir, 'api.py'))) {
```
Comment (alexcasalboni) [inline src/config.js]: Here the assumption is that we have either JS or Python functions, right? Isn't there an easy way to detect and parse both runtimes?

**Comment by alexcasalboni** [comment]
@lusentis please have a look at the last few commits! I didn't have time to properly test the devInstrument for Python, but it should be a good starting point :)

**Comment by lusentis** [comment]
Thank you! I'll have a look this evening.

File: src/config.js (hunk 13 above)
Comment (lusentis) [inline src/config.js]: Yes, currently.
There are a few things we need to figure out before allowing multiple runtimes at once. After we've merged this branch into master I'll open a PR to discuss :)

**Comment by lusentis** [comment]
(my Mac broke and I can't get this to work on Windows..., I'll need a few more days to review this)

**Comment by alexcasalboni** [comment]
@lusentis do you have a new Mac now? ;)

**Comment by lusentis** [comment]
No, it's a PC 💻. I promise I'll work on this in the next days.

**Pull request closed without merging**